cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc

```diff
@@ -709,94 +709,96 @@ year_month_day GetFlooredYmd(int64_t arg, int multiple, Localizer localizer_) {
 
 template <typename Duration, typename Unit, typename Localizer>
 const Duration FloorTimePoint(const int64_t arg, const int64_t multiple,
-                              Localizer localizer_, Status* st) {
+                              Localizer localizer_, const RoundTemporalOptions options,
+                              Status* st) {
   const auto t = localizer_.template ConvertTimePoint<Duration>(arg);
   const Unit d = floor<Unit>(t).time_since_epoch();
 
   if (multiple == 1) {
     return localizer_.template ConvertLocalToSys<Duration>(duration_cast<Duration>(d),
-                                                           st);
+                                                           options, st);
   } else {
     const Unit unit = Unit{multiple};
     const Unit m =
         (d.count() >= 0) ? d / unit * unit : (d - unit + Unit{1}) / unit * unit;
     return localizer_.template ConvertLocalToSys<Duration>(duration_cast<Duration>(m),
-                                                           st);
+                                                           options, st);
   }
 }
 
 template <typename Duration, typename Localizer>
 const Duration FloorWeekTimePoint(const int64_t arg, const int64_t multiple,
                                   Localizer localizer_, const Duration weekday_offset,
-                                  Status* st) {
+                                  const RoundTemporalOptions options, Status* st) {
   const auto t = localizer_.template ConvertTimePoint<Duration>(arg) + weekday_offset;
   const weeks d = floor<weeks>(t).time_since_epoch();
 
   if (multiple == 1) {
     return localizer_.template ConvertLocalToSys<Duration>(duration_cast<Duration>(d),
-                                                           st) -
+                                                           options, st) -
            weekday_offset;
   } else {
     const weeks unit = weeks{multiple};
     const weeks m =
         (d.count() >= 0) ? d / unit * unit : (d - unit + weeks{1}) / unit * unit;
     return localizer_.template ConvertLocalToSys<Duration>(duration_cast<Duration>(m),
-                                                           st) -
+                                                           options, st) -
            weekday_offset;
   }
 }
 
 template <typename Duration, typename Unit, typename Localizer>
 Duration CeilTimePoint(const int64_t arg, const int64_t multiple, Localizer localizer_,
-                       Status* st) {
+                       const RoundTemporalOptions options, Status* st) {
```

Comment:
Would it make sense to have `arrow_vendored::date::choose` as a template parameter rather than passing options every time the kernel is called? (I'm not certain just asking)

Also note that we could probably do more templating for the [ceil/floor/round kernels](https://issues.apache.org/jira/browse/ARROW-15787), but that's out of scope here.